### ISSUE TITLE
Compile Error (OS X)

### ISSUE THREAD

**Issue by alzmcr** [dmlc/cxxnet#70]
I've been trying to compile cxxnet for quite a while on OSX, trying every now and then ... trying to fix problems along the way looking on Kaggle forum and other places, but today I'm stuck with this error which I couldn't find any explanations - hence I'm posting!

![screen shot 2015-03-25 at 12 45 57](https://cloud.githubusercontent.com/assets/4122640/6824768/f54f1dfc-d2ec-11e4-87bd-15025880ce09.png)

Any idea why I'm getting "CLOCK_REALTIME" and "clock_gettime" out of scope?

Here is how my config looks like:

export CC = gcc-4.9
export CXX = g++-4.9
export NVCC = nvcc

USE_CUDA = 1
USE_CUDA_PATH = /Developer/NVIDIA/CUDA-7.0

USE_OPENCV = 1 
USE_OPENCV_DECODER = 1
USE_CUDNN = 0
USE_CUDNN_PATH = NONE

USE_STATIC_MKL = NONE
USE_BLAS = blas
USE_INTEL_PATH = NONE

USE_DIST_PS = 0
PS_PATH = NONE
PS_THIRD_PATH = NONE

ADD_LDFLAGS = -ljpeg

ADD_CFLAGS =

**Comment by tqchen** [comment]
I pushed a fix for the nan problem. For timer, it should have went to another direction in 

https://github.com/dmlc/cxxnet/blob/master/src/utils/timer.h#L18

**Comment by alzmcr** [comment]
I've just tried with the new push and indeed the nan problem is solved. the clock still remain though .. my C experience is very close to zero, so once I've figured out how that "ifdef" works I'll try a manual change. If is something quick to change a "share" will be much appreciated! btw if you can think of any reason why this is happening I'm happy to try!

**Comment by tqchen** [comment]
It was likely due to the fact that __MACH__ was not defined in your compiler. Maybe you can try to change it to __APPLE__ and see if things work for you

Thanks

**Comment by alzmcr** [comment]
I've tried with __APPLE__, __APPLE_CC__, __OSX__ ... nothing :(

**Comment by tqchen** [comment]
you want to keep the two underscores so it should be __APPLE__

**Comment by alzmcr** [comment]
hehe, they all have the underscores... that's why is bold! I don't know the markup to prevent going bold, sorry :)

I've tried adding #define __MACH__ (with the underscores) but I still got the same error :(

**Comment by tqchen** [comment]
alrite, it turns out to be our problem introduced in recent windows support. I have pushed a fix, please see if it works for you now.

Sorry about the inconvenience

**Comment by alzmcr** [comment]
yey - I now got other error, I'll see if I can fix it by myself -  thanks heaps Tianqi!!!

**Comment by pchankh** [comment]
I cannot seem to compile for OSX too.  Was able to compile for version 1 of cxxnet.

**Comment by zhengsx** [comment]
Hi Tianqi, I got this error.

g++ -DMSHADOW_FORCE_STREAM -Wall -g -O3 -I./mshadow/  -fPIC -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CUDA=0 -DMSHADOW_DIST_PS=0 -DCXXNET_USE_OPENCV=1 -DCXXNET_USE_OPENCV_DECODER=1   -o bin/cxxnet src/local_main.cpp layer_cpu.o updater_cpu.o nnet_cpu.o data.o main.o nnet_ps_server.o -pthread -lm -Wl,--start-group /opt/intel/mkl/lib/libmkl_intel_lp64.a /opt/intel/mkl/lib/libmkl_core.a /opt/intel/mkl/lib/libmkl_intel_thread.a -Wl,--end-group -liomp5 -ldl -lpthread -lm `pkg-config --libs opencv` -ljpeg
ld: unknown option: --start-group
collect2: error: ld returned 1 exit status

**Comment by tqchen** [comment]
We will look into it. In the meanwhile, you can use other BLAS such as openBLAS or ATLAS, where the linking problem is easier

**Comment by zhengsx** [comment]
yeah I try openBlas, and then I stuck with this error. It seems linkers can't find the implementations of some functions?
{
 Undefined symbols for architecture x86_64:
  "std::string::compare(char const*) const", referenced from:
              cxxnet::layer::FixConnectLayer<mshadow::cpu>::InitConnection(std::vector<cxxnet::layer::Node<mshadow::cpu>*, std::allocator<cxxnet::layer::Node<mshadow::cpu>*> > const&, std::vector<cxxnet::layer::Node<mshadow::cpu>*, std::allocator<cxxnet::layer::Node<mshadow::cpu>*> > const&, cxxnet::layer::ConnectState<mshadow::cpu>*) in layer_cpu.o
      cxxnet::updater::AsyncUpdater<mshadow::cpu>::SetParam(char const*, char const*) in updater_cpu.o
      cxxnet::utils::MetricSet::Print(char const*) in nnet_cpu.o
...
      cxxnet::nnet::NeuralNetThread<mshadow::cpu>::ThreadEntry(void*) in nnet_cpu.o
      cxxnet::nnet::CXXNetUpdater::CXXNetUpdater() in nnet_ps_server.o
      cxxnet::nnet::CXXNetUpdater::InitUpdater(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in nnet_ps_server.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [bin/cxxnet] Error 1
}

sorry but looking forward ur reply
